### PR TITLE
ActivityPub handling

### PR DESCRIPTION
**Pull request by srosset81**
Doit être mergé avec https://github.com/YesWiki/yeswiki/pull/406

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -25,6 +25,8 @@ To integrate the webhook with a Mattermost chatroom, follow this guideline:
 - Create a webhook with the configurations you wish to use
 - On the YesWiki BazaR page, choose "Mattermost" and copy the URL provided by Mattermost
 
+You can define `webhooks_bot_name` and `webhooks_bot_icon` in `wakka.config.php` if you want to change the look of the bot which will post on Mattermost (you must allow this in the Mattermost configurations).
+
```
Comment (mrflos) [inline README.md:29]: Nice! I was fed up being the name of the bot

File: README.md [hunk 2]
```diff
@@ -44,6 +46,51 @@ The data will be sent formatted like this:
...
+```json
+{
+  "@context": "https://www.w3.org/ns/activitystreams",
+  "type":"Create",
+  "to": [
+    "http://localhost/?WikiAdmin/followers",
+    "https://www.w3.org/ns/activitystreams#Public"
+  ],
+  "actor": "http://localhost/?WikiAdmin",
```
Comment (mrflos) [inline README.md:63]: How anonymous contributions would be handled ?

Comment (mrflos) [reply README.md:63]: ok just saw documentation more down

File: README.md [hunk 3]
```diff
@@ -44,6 +46,51 @@ The data will be sent formatted like this:
...
+
+#### ActivityPub actors
+By default, the actor posting the activity will be the logged-in user. You can also mark a field with the context `https://www.w3.org/ns/activitystreams#author` to override this value.
+
+You can set two parameters in the `wakka.config.php` file :
+
+- `webhooks_activitypub_default_actor`: The URL you associate with this parameter will be used as an actor for **all** ActivityPub activities.
+
+- `webhooks_activitypub_actors_base_url`: The base URL to use for all actors. For example if you use `https://activitypub.server/actor/`, then the URI of user WikiAdmin will be `https://activitypub.server/actor/WikiAdmin` instead of `http://localhost/?WikiAdmin`.
+
```
Comment (mrflos) [inline README.md:92]: Maybe we can consider an actor called Anonymous that would be author of all non identified contributions?

File: libs/functions.php [hunk 4]
```diff
@@ -45,16 +45,85 @@ function get_notification_text($data, $action_type, $user_name)
...
+    return $date_array[0] . "T" . $date_array[1] . "Z";
+}
+
+function get_actor_uri($actor) {
+    if ($GLOBALS['wiki']->config['webhooks_activitypub_default_actor'] !== '') {
+        // If a default global-wide actor is defined, use it
+        return $GLOBALS['wiki']->config['webhooks_activitypub_default_actor'];
+    } else {
+        // If no field is marked as an actor, take the current logged-in user
+        if( !$actor ) $actor = $GLOBALS['wiki']->href('', $GLOBALS['wiki']->getUser() ? $GLOBALS['wiki']->getUser()['name'] : _t('WEBHOOKS_ANONYMOUS_USER'));
```
Comment (mrflos) [inline libs/functions.php:59]: Ok you seem to have a anonymous user

File: libs/functions.php [hunk 5]
```diff
@@ -45,16 +45,85 @@ function get_notification_text($data, $action_type, $user_name)
...
+                "actor" => $actorUri,
+                "object" => $object
+            ];
+
         case WEBHOOKS_FORMAT_MATTERMOST:
             return [
-                "username" => $GLOBALS['wiki']->config['WEBHOOKS_BOT_NAME'],
-                "icon_url" => $GLOBALS['wiki']->config['WEBHOOKS_BOT_ICON'],
+                "username" => $GLOBALS['wiki']->config['webhooks_bot_name'],
+                "icon_url" => $GLOBALS['wiki']->config['webhooks_bot_icon'],
```
Comment (mrflos) [inline libs/functions.php:126]: ❤️

File: wiki.php [hunk 6]
```diff
@@ -18,19 +19,28 @@
...
+// If an actor is defined, we can override the base URL with this config
+$wakkaConfig['webhooks_activitypub_actors_base_url'] = getConfigValue('webhooks_activitypub_actors_base_url', '', $wakkaConfig);
+
 // Bot config (works for Mattermost if username and profile picture override is enabled)
 // See https://docs.mattermost.com/developer/webhooks-incoming.html
-$wakkaConfig['WEBHOOKS_BOT_NAME'] = getConfigValue('WEBHOOKS_BOT_NAME', "YesWiki Bot", $wakkaConfig);
-$wakkaConfig['WEBHOOKS_BOT_ICON'] = getConfigValue('WEBHOOKS_BOT_ICON', "https://yeswiki.net/files/PageHeader_yeswikiprovisoire_vignette_97_97_20181206153605_20181206154004.png", $wakkaConfig);
\ No newline at end of file
+$wakkaConfig['webhooks_bot_name'] = getConfigValue('webhooks_bot_name', "YesWiki Bot", $wakkaConfig);
+$wakkaConfig['webhooks_bot_icon'] = getConfigValue('webhooks_bot_icon', "https://yeswiki.net/files/PageHeader_yeswikiprovisoire_vignette_97_97_20181206153605_20181206154004.png", $wakkaConfig);
```
Comment (mrflos) [inline wiki.php]: It's a detail, but maybe use local, relative url like `tools/webhooks/presentation/images/default-bot.png` with a image like https://pixabay.com/illustrations/bot-icon-robot-automated-cyborg-2883144/

**Review by mrflos — APPROVED** [review]
Seems ok, just a few discussions opened, great job @srosset81 !

File: wiki.php (hunk 6 above)
Comment (srosset81) [reply wiki.php]: Done :)

**Review by mrflos — APPROVED** [review]

**Pull request merged**